File: akka-samples/scala-samples/src/main/scala/actors/supervision/SupervisingActor.scala
package actors.supervision

import akka.actor.typed.SupervisorStrategy
import akka.actor.typed.scaladsl.Behaviors
import akka.actor.typed.{ActorSystem, Behavior}

import scala.concurrent.duration.DurationInt

object SupervisingActor {
  def main(args: Array[String]): Unit = {
    ActorSystem[String](Guardian(), "supervision")
  }
}

object Guardian {
  def apply(): Behavior[String] =
    Behaviors.supervise(behave())
      .onFailure[IllegalStateException](SupervisorStrategy.restart.withLimit(3, 3.seconds))

  def behave(): Behavior[String] =
    Behaviors.setup[String] { context =>
      println("Starting Actor...")
      throw new IllegalStateException("something went wrong")

      println("Actor started!") // unreachable
      Behaviors.stopped
    }
}

File: akka-samples/java-samples/src/main/java/actors/supervision/SupervisingActor.java
package actors.supervision;

import akka.actor.typed.ActorSystem;
import akka.actor.typed.Behavior;
import akka.actor.typed.SupervisorStrategy;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;

import java.time.Duration;

public class SupervisingActor extends AbstractBehavior<String> {

    public SupervisingActor(ActorContext<String> context) {
        super(context);
    }

    public static Behavior<String> create() {
        return Behaviors.supervise(SupervisedActor.create())
                .onFailure(IllegalStateException.class, SupervisorStrategy.restart().withLimit(3, Duration.ofSeconds(3)));
    }

    public static void main(String[] args) {
        ActorSystem.create(SupervisingActor.create(), "supervision");
    }

    @Override
    public Receive<String> createReceive() {
        return newReceiveBuilder().build();
    }
}
